File: simnibs/examples/group_average_fsavg.m
% This example wil go throgh simulations and calculate
% the average and the standard deviation of the normal component
% of the electric field in FsAverage space
% 
% It is a follow-up to the "run_simulations" example

%% Load simulation results
subjects = {'sub01', 'sub09', 'sub10', 'sub12', 'sub15'};
results_folder = fullfile('bipolar', 'fsavg_overlays');
fsavg_msh_name = '_TDCS_1_scalar_fsavg.msh';
field_name = 'E_normal';

fields = {};
for i = 1:length(subjects)
    sub = subjects{i};
    % load mesh with results transformed to fsaverage space
    m = mesh_load_gmsh4(fullfile(pwd, sub, results_folder, [sub fsavg_msh_name]));
    % Save the field of each subject
    fields{i} = m.node_data{get_field_idx(m, field_name, 'node')}.data;
end
%% Calculate and plot averages
% Calculate
fields = cell2mat(fields);
avg_field = mean(fields, 2);
std_field = std(fields, 0, 2);
% Plot
m.node_data = {}; %cleanup fields
m.node_data{1}.data = avg_field; % add average field
m.node_data{1}.name = [field_name '_avg'];
m.node_data{2}.data = std_field; % add std field
m.node_data{2}.name = [field_name '_std'];

% show surfaces with fields
mesh_show_surface(m, 'field_idx', [field_name '_avg'])
mesh_show_surface(m, 'field_idx', [field_name '_std'])

%% Calculate average in an ROI defined using an atlas
% load atlas and define a region
[m, snames]=mesh_load_fssurf('fsaverage','label','a2009s');
region_name = 'lh.S_precentral-sup-part';
roi_idx=find(strcmpi(snames, region_name));
node_idx = m.node_data{end}.data==roi_idx;
% visualize region
m.node_data{end+1}.data = int8(node_idx);
m.node_data{end}.name = region_name;
mesh_show_surface(m, 'field_idx', region_name)

% calculate a weighted mean using the node areas
nodes_areas = mesh_get_node_areas(m);
avg_field_roi = sum(avg_field(node_idx).*nodes_areas(node_idx))/sum(nodes_areas(node_idx));
fprintf('Average %s in %s: %f\n', field_name, region_name, avg_field_roi)

File: simnibs/examples/analysis/group_average_fsavg.py
'''
    This example wil go throgh simulations and calculate
    the average and the standard deviation of the normal component
    of the electric field in FsAverage space

    It is a follow-up to the "run_simulations" example
'''
import os
import numpy as np
import simnibs

## Load simulation results
subjects = ['sub01', 'sub09', 'sub10', 'sub12', 'sub15']
results_folder = 'fsavg_overlays'
fsavg_msh_name = '_TDCS_1_scalar_fsavg.msh'
field_name = 'E_normal'

fields = []
for sub in subjects:
    # read mesh with results transformed to fsaverage space
    results_fsavg = simnibs.read_msh(
        os.path.join('bipolar', sub, results_folder, sub + fsavg_msh_name)
    )
    # save the field in each subject
    fields.append(results_fsavg.field[field_name].value)

## Calculate and plot averages
# Calculate
fields = np.vstack(fields)
avg_field = np.mean(fields, axis=0)
std_field = np.std(fields, axis=0)

# Plot
results_fsavg.nodedata = [] # cleanup fields
results_fsavg.add_node_field(avg_field, 'E_normal_avg') # add average field
results_fsavg.add_node_field(std_field, 'E_normal_std') # add std field

# show surface with the fields
results_fsavg.view(visible_fields='E_normal_avg').show()

## Calculate average in an ROI defined using an atlas
# load atlas and define a region
atlas = simnibs.get_atlas('HCP_MMP1')
region_name = 'lh.4'
roi = atlas[region_name]
# visualize region
results_fsavg.add_node_field(roi, region_name)
results_fsavg.view(visible_fields=region_name).show()

# calculate mean field using a weighted mean
node_areas = results_fsavg.nodes_areas()
avg_field_roi = np.average(avg_field[roi], weights=node_areas[roi])
print(f'Average {field_name} in {region_name}: ', avg_field_roi)
results_fsavg.add_node_field(roi, region_name)
